src/rooted_gradient_transforms.cpp

```diff
@@ -129,44 +133,94 @@ double UpdateHeightParameterGradientUnweightedLogDensity(
     const RootedTree &tree, const std::vector<double> &height_gradient) {
   size_t leaf_count = tree.LeafCount();
   size_t root_id = tree.Topology()->Id();
 
   // Calculate node ratio gradient
-  std::vector<double> gradientLogDensity =
+  std::vector<double> gradient_log_density =
       UpdateGradientUnWeightedLogDensity(tree, height_gradient);
 
   // Calculate root height gradient
-  gradientLogDensity[root_id - leaf_count] =
+  gradient_log_density[root_id - leaf_count] =
       UpdateHeightParameterGradientUnweightedLogDensity(tree, height_gradient);
 
-  // Add gradient of log Jacobian determinant
-  std::vector<double> log_time = GetLogTimeArray(tree);
+  return gradient_log_density;
+}
 
-  std::vector<double> gradientLogJacobianDeterminant =
-      UpdateGradientUnWeightedLogDensity(tree, log_time);
-  gradientLogJacobianDeterminant[root_id - leaf_count] =
-      UpdateHeightParameterGradientUnweightedLogDensity(tree, log_time);
+std::vector<double> RatioGradientOfBranchGradient(
+    const RootedTree &tree, const std::vector<double> &branch_gradient) {
+  size_t leaf_count = tree.LeafCount();
+  size_t root_id = tree.Topology()->Id();
 
-  for (int i = 0; i < gradientLogDensity.size() - 1; i++) {
-    gradientLogDensity[i] +=
-        gradientLogJacobianDeterminant[i] - 1.0 / tree.height_ratios_[i];
-  }
+  // Calculate node height gradient
+  std::vector<double> height_gradient = HeightGradient(tree, branch_gradient);
 
-  gradientLogDensity[root_id - leaf_count] +=
-      gradientLogJacobianDeterminant[root_id - leaf_count];
+  // Calculate ratios and root height gradient
+  std::vector<double> gradient_log_density =
+      RatioGradientOfHeightGradient(tree, height_gradient);
 
-  return gradientLogDensity;
+  // Calculate gradient of log Jacobian determinant
+  std::vector<double> gradient_log_jacobian_determinant =
+      GradientLogDeterminantJacobian(tree);
+
+  for (int i = 0; i < gradient_log_jacobian_determinant.size() - 1; i++) {
+    gradient_log_density[i] += gradient_log_jacobian_determinant[i];
+  }
+  gradient_log_density[root_id - leaf_count] +=
+      gradient_log_jacobian_determinant[root_id - leaf_count];
+
+  return gradient_log_density;
 }
 
 std::vector<double> RatioGradientOfBranchGradient(
-    const RootedTree &tree, const std::vector<double> &branch_gradient) {
+    const RootedTree &tree, const std::vector<double> &branch_gradient,
+    const std::optional<PhyloFlags> flags /*= std::nullopt*/) {
+  size_t leaf_count = tree.LeafCount();
+  size_t root_id = tree.Topology()->Id();
+
   // Calculate node height gradient
   std::vector<double> height_gradient = HeightGradient(tree, branch_gradient);
 
-  return RatioGradientOfHeightGradient(tree, height_gradient);
+  // Calculate ratios and root height gradient
+  std::vector<double> gradient_log_density =
+      RatioGradientOfHeightGradient(tree, height_gradient);
+
+  // NOTE: Here I added them so tests do not fail but that's what we want to avoid,
```

Comment:
I can't say I get this comment... the code seems sensible.

Reply:
Old notes from Fourment.  Removed.

Reply:
Added descriptions.